pyzx/simplify.py

```diff
@@ -203,74 +231,216 @@ def full_reduce(g: BaseGraph[VT,ET], quiet:bool=True, stats:Optional[Stats]=None
-                if (p1+p2).denominator <= 2:
-                    del self.phantom_phases[v1]
-                v1,i1 = v3,i3
-                m1 = cast(Literal[1, -1], m1*self.simplifygraph.phase_mult[i3])
-            else:
-                self.phantom_phases[v1] = (v2,i2)
-                self.simplifygraph.phase_mult[i2] = m2
-                return
-        if p1.denominator <= 2 or p2.denominator <= 2: raise Exception("Clifford phases here??")
-        # Both have non-Clifford phase
-        if m1*m2 == 1: phase = (p1 + p2)%2
-        else: phase = p1 - p2
-        self.mastergraph.set_phase(v1,phase)
-        self.mastergraph.set_phase(v2,0)
-
-        self.simplifygraph.phase_mult[i2] = 1
-
-    def full_reduce(self, quiet:bool=True, stats:Optional[Stats]=None) -> None:
-        full_reduce(self.simplifygraph,quiet=quiet, stats=stats)
-
-
-
-def to_gh(g: BaseGraph[VT,ET],quiet:bool=True) -> None:
+        self.original_graph: BaseGraph[VT,ET] = g.copy()
+        self.parent_vertex: Dict[VT,VT] = {}
+        self.vertex_rank: Dict[VT,int] = {}
+        self.phase_mult: Dict[VT,int] = {}
+        self.non_clifford_vertices: Set[VT] = set()
+        for v in self.original_graph.vertices():
+            if self.original_graph.phase(v).denominator > 2:
+                self.parent_vertex[v] = v
+                self.vertex_rank[v] = 0
+                self.phase_mult[v] = 1
+                self.non_clifford_vertices.add(v)
+    
+    def parent(self, v: VT) -> VT:
+        if self.parent_vertex[v] != v:
+            self.parent_vertex[v] = self.parent(self.parent_vertex[v])
+        return self.parent_vertex[v]
+    
+    def get_vertex_groups(self) -> List[List[VT]]:
+        vertex_groups: defaultdict[VT,List[VT]] = defaultdict(list)
+        for v in self.non_clifford_vertices:
+            root = self.parent(v)
+            vertex_groups[root].append(v)
+        return list(vertex_groups.values())
+    
+    def fuse_phases(self, v1: VT, v2: VT) -> None:
+        if not all(v in self.non_clifford_vertices for v in (v1, v2)): return
+        self.parent_vertex[v2] = v1
+        self.vertex_rank[v2] = self.vertex_rank[v1]
+        self.vertex_rank[v1] += 1
+    
+    def phase_negate(self, v: VT) -> None:
+        root = self.parent(v)
+        verts = [vert for vert in self.non_clifford_vertices if self.parent(vert) == root]
+        for vert in verts:
+            self.phase_mult[vert] *= -1
+    
+    def init_simplify_graph(self, fusing_mode: bool = True) -> None:
+        self.simplify_graph = self.original_graph.clone()
+        self.simplify_graph.set_phase_teleporter(self, fusing_mode)
+    
+    def teleport_phases(self, store:bool = False) -> None:
+        self.init_simplify_graph()
+        full_reduce(self.simplify_graph)
+        self.init_simplify_graph(fusing_mode = False)
+        if not store: self.simplify_graph.place_tracked_phases()
+
+def teleport_reduce(g: BaseGraph[VT,ET], store: bool = False) -> None:
```

Comment:
I changed it to return None to align with all the other functions in `simplify.py ` so I'll update the tests instead, unless you would rather maintain returning the graph to avoid confusion?

Reply:
I think it makes sense to align it with the other functions and update the tests instead.

Does @jvdwetering have any opinions about the return value?